pkg/beaconfqdn/mongodb.go

```diff
@@ -151,8 +178,189 @@ func (r *repo) Upsert(hostnameMap map[string]*hostname.Input, minTimestamp, maxT
 		bar.IncrBy(1)
 
 	}
+
 	p.Wait()
 
 	// start the closing cascade (this will also close the other channels)
 	dissectorWorker.close()
 }
+
+// affectedHostnameIPs gathers all of the hostnames associated with the external IPs which generated
+// traffic in this run. Each hostname entry is returned along with its list of associated resolved IPs.
+func (r *repo) affectedHostnameIPs(hostMap map[string]*host.Input) ([]hostnameIPs, error) {
+
+	// We can only submit around 200,000 hosts in a single query to MongoDB due to the 16MB document limit.
+	// 1 Megabyte / Size of Unique IP BSON query ~= 200,000
+	// 16 * 1024 * 1024 / (64 for IPv4, 83 for IPv6) = (262144 for IPv4, 202135 for IPv6)
+
+	// provide a fast path for datasets with less than 200,000 hosts
+	if len(hostMap) <= 200000 {
+		return r.affectedHostnameIPsSimple(hostMap)
+	}
+	return r.affectedHostnameIPsChunked(hostMap)
+}
+
+// affectedHostnameIPsSimple implements affectedHostnameIPs but should only be called with hostMaps with
+// roughly less than 200,000 external hosts.
+func (r *repo) affectedHostnameIPsSimple(hostMap map[string]*host.Input) ([]hostnameIPs, error) {
+	// preallocate externalHosts slice assuming at least half of the observed hosts are external
+	var externalHosts []data.UniqueIP = make([]data.UniqueIP, 0, len(hostMap)/2)
```

Comment:
The only disadvantage is RAM usage as you said. 

I went with `/2` because I took a guess at how slice allocation works and assumed that it doubled the size of the backing array when reallocation happens. It looks like that Go does actually double the size of the underlying slice when you `append` more items than the slice has capacity fore: https://github.com/golang/go/blob/0a7f00ae239570d664488085f9c395919bc69066/src/runtime/slice.go#L164. 

So, with `/2`, we guarantee that we will only need to re-allocate once in the worst case. I could add that in as a comment if you'd like.

I'm happy to change it to `len(hostMap)` however if you feel that best suits our needs.

Reply:
Oh nope, if there is just one re-allocation then I think it's worth it to go minimal. I was just concerned if it was going to be re-allocating with every new entry. This way is great though if it just does the one.

Reply:
Chill, I'll add the explanation above as a comment.